### PR TITLE
added solution

### PR DESCRIPTION
**Pull request by Rudiachenko**

### CODE REVIEW

**Review by andromash — APPROVED** [review]
Good job!

File: src/main/java/com/dev/cinema/util/HashUtil.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        StringBuilder hashedPassword = new StringBuilder();
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance(ALGORITHM);
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            for (byte part : digest) {
+                hashedPassword.append(String.format("%02x", part));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("No cryptographic algorithm is available.");
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/util/HashUtil.java]: let's add e to exception message

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.Optional;
+
+@Service
+public class UserServiceImpl implements UserService {
+    @Inject
+    private UserDao userDao;
+
+    @Override
+    public User add(User user) {
+        byte[] salt = HashUtil.getSalt();
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]: you don't need separate variable for salt, set it to user right away

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        String password = HashUtil.hashPassword(user.getPassword(), salt);
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]:
```suggestion
        String hashedPassword = HashUtil.hashPassword(user.getPassword(), salt);
```

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        }
+        throw new AuthenticationException("You entered incorrect email or password.");
+    }
+
+    @Override
+    public User register(String email, String password) {
+        return userService.add(new User(email, password));
+    }
+
+    private boolean isValid(String password, User user) {
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]:
```suggestion
    private boolean isPasswordValid(String password, User user) {
```

File: src/main/java/com/dev/cinema/security/AuthenticationService.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.dev.cinema.security;
+
+import com.dev.cinema.model.User;
+import exceptions.AuthenticationException;
+
+public interface AuthenticationService {
+    User login(String email, String password) throws AuthenticationException;
+
+    /**
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/security/AuthenticationService.java]: we don't need this comment

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**